# UKCI2016-Edgetics
The R work described in our conference paper presented at UKCI-2016 in Lancaster, 7th-9th Sept. If you find this code useful please cite our work:

K. McGarry, K. Emery, V. Varnakulasingam, S. McDonald and M. Ashton . Complex network based computational techniques for edgetic modelling of mutations implicated with human diseases. The 16th UK Workshop on Computational Intelligence, UKCI-2016, Lancaster, Uk, 7th-9th Sept, 2016.

Complex networks are a graph theoretic method that can model genetic mutations, in particular single nucleotide polymorphisms (snp’s) which are genetic variations that only occur at single position in a DNA sequence. These can potentially cause the amino acids to be changed and may affect protein function and thus structural stability which can contribute to developing diseases. We show how snp’s can be represented by complex graph structures, the connectivity patterns if represented by graphs can be related to human diseases,  where the proteins are the nodes (vertices) and the interactions between them are represented by links (edges). Disruptions caused by mutations can be explained as loss of connectivity such as the deletion of nodes or edges in the network (hence the term {\it edgetics}). Furthermore, diseases appear to be interlinked with hub genes causing multiple problems and this has led to the concept of the human disease network or {\it diseasome}. Edgetics is a relatively new concept which is proving effective for modelling the relationships between genes, diseases and drugs which were previously considered intractable problems.
